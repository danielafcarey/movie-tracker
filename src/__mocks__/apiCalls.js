import { mockMovies } from '../mockData';

export const fetchMovies = jest.fn().mockImplementation( () => {
  return Promise.resolve( mockMovies );
});
