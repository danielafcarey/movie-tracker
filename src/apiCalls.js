import { apiKey } from './apiKey';

const fetchMovies = async () => {
  const url = `https://api.themoviedb.org/3/movie/now_playing?api_key=${apiKey}`;
  try {
    const response = await fetch(url);
    if (response.status === 200) {
      const data = await response.json();
      return data;
    } else {
      throw Error(`Status failure: ${response.status}`);
    }
  } catch (error) {
    throw error;
  }
}; 

const fetchUsers = async () => {
  const url = 'http://localhost:3000/api/users';
  try {
    const response = await fetch(url);
    if (response.status === 200) {
      const data = await response.json();
      return data.data;
    } else {
      throw Error(response.status);
    }
  } catch (error) {
    throw Error(error);
  }

};

const fetchFavorites = async (id) => {
  const url = `http://localhost:3000/api/users/${id}/favorites`;
  try {
    const response = await fetch(url);
    if (response.status === 200) {
      const data = await response.json();
      return data.data;
    } else {
      throw Error(response.status);
    }
  } catch (error) {
    throw Error(error);
  }
}

const postFavorite = async (userId, movie) => {
  const url = `http://localhost:3000/api/users/favorites/new`;
  const favorite = Object.assign({}, movie, { user_id: userId, overview: '' });
  const optionsObject = {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(favorite)
  };

  try {
    const response = await fetch(url, optionsObject);  
    if (response.status !== 200) {
      throw Error(response.status);
    }
  } catch (error) {
    throw Error(error);
  }
}

const deleteFavorite = async (userId, movieId) => {
  const url = `http://localhost:3000/api/users/${userId}/favorites/${movieId}`; 
  try {
    const response = await fetch(url);
    if (response.status !== 200) {
      throw Error(response.status)
    }
  } catch(error) {
    throw Error(error);
  }
}

export {
  fetchMovies,
  fetchUsers,
  fetchFavorites,
  postFavorite,
  deleteFavorite
};

