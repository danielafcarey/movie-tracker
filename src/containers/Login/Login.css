.login {
  display: flex;
  justify-content: center;
  align-items: center; }
  .login form {
    box-shadow: 0px 0px 20px 0px black;
    border-radius: 5px;
    margin-top: 30px;
    padding: 20px;
    text-align: center; }
    .login form input {
      display: block;
      width: 300px;
      font-size: 20px;
      margin-top: 30px;
      padding: 3px; }
      .login form input::placeholder {
        padding: 3px; }
    .login form p {
      color: #b33533;
      margin: 3px; }
    .login form button {
      width: 305px;
      background: #b33533;
      color: white;
      font-size: 20px;
      font-weight: 400;
      margin-bottom: 30px;
      margin-top: 30px;
      padding: 5px;
      border: none; }
      .login form button:hover {
        cursor: pointer;
        opacity: 0.8; }
