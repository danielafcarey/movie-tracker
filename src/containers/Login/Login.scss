.login {
  display: flex;
  justify-content: center;
  align-items: center;

  form {
    box-shadow: 0px 0px 20px 0px black;
    border-radius: 5px;
    margin-top: 30px;
    padding: 20px;
    text-align: center;

    input {
      display: block;
      width: 300px;
      font-size: 20px;
      margin-top: 30px;
      padding: 3px;

      &::placeholder {
        padding: 3px;
      }
    }

    p {
      color: #b33533;
      margin: 3px;
    }

    button {
      width: 305px;
      background: #b33533;
      color: white;
      font-size: 20px;
      font-weight: 400;
      margin-bottom: 30px;
      margin-top: 30px;
      padding: 5px;
      border: none;

      &:hover {
        cursor: pointer;
        opacity: 0.8;
      }
    }

  }
}
