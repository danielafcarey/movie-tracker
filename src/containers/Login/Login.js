import React, { Component } from 'react';
import { fetchUsers, fetchFavorites } from '../../apiCalls';
import { updateCurrentUser, updateFavorites } from '../../actions';
import { connect } from 'react-redux'

class Login extends Component {
  constructor(props){
    super(props);

    this.state = {
      email: '',
      password: '',
      id: null
    }
  }

  handleChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value })
  }

  verifyUser = async () => {
    const { email, password } = this.state;
    const users = await fetchUsers();
    const userMatch = users.find(user => {
      if(user.password === password && user.email === email) {
        return user
      }
    })
    if (userMatch) {
      this.setState({ id: userMatch.id })
      return userMatch.id
    } else {
      return undefined;
    }
  }

  handleSubmit = async (event) => {
    event.preventDefault();
    const userId = this.verifyUser();
    if (userId) {
      this.props.updateCurrentUser(userId);
      const favorites = await fetchFavorites(userId)
    } else {
      alert('User Does Not Exist')
    }
  }

  render() {
    return(
      <form
        type='submit'
        onSubmit={this.handleSubmit}
      >
        <input 
          type='text'
          placeholder='Email'
          name='email'
          value={this.state.email}
          onChange={this.handleChange}
        />
        <input
          type='password'
          placeholder='Password'
          name='password'
          value={this.state.password}
          onChange={this.handleChange}          
        />
        <button>Log In</button>
      </form>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  updateCurrentUser: (id) => dispatch(updateCurrentUser(id)),
  updateFavorites: (favorites) => dispatch(updateFavorites(favorites)),
})

export {
  Login,
  mapDispatchToProps
};

export default connect(null, mapDispatchToProps)(Login)