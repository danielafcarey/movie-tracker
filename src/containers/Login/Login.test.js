import React from 'react';
import { shallow } from 'enzyme';
import { Login } from './Login';
import * as apiCalls from '../../apiCalls'

describe('Login', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Login />);
  });

  it('matches the snapshot', () => {
    // expect(wrapper).toMatchSnapshot();
  });

  it('has a default state of email and password', () => {

    expect(wrapper.state('email')).toEqual('')
    expect(wrapper.state('password')).toEqual('')
  })

  it('changes email in state onChange', () => {
    const mockEvent = { target: { value: 'garbage', name: 'email'}}
    wrapper.instance().handleChange(mockEvent)

    expect(wrapper.state('email')).toEqual(mockEvent.target.value)
  })

  it('changes password in state onChange', () => {
    const mockEvent = { target: { value: 'garbage', name: 'password' } }
    wrapper.instance().handleChange(mockEvent)

    expect(wrapper.state('password')).toEqual(mockEvent.target.value)
  })

  describe('verifyUser', () => {

    beforeEach(() => {
      
      apiCalls.fetchUsers = jest.fn().mockImplementation(() => {
        return [
          {
            id: 1,
            email: 'garbageman@gmail.com',
            password: 'ismellbad',
            name: 'jerry'
          }
        ]
      })

      const mockState = {
        email: 'garbageman@gmail.com',
        password: 'ismellbad'
      }

      wrapper.setState(mockState);
    })

    it('calls fetchUsers', async () => {
      
      wrapper.instance().verifyUser()

      expect(apiCalls.fetchUsers).toHaveBeenCalled();
    })

    it('returns userId if there is a match', async () => {
      
      const result = await wrapper.instance().verifyUser();
      const expected = 1;

      expect(result).toEqual(expected)
    })

    it('returns undefined if there is no match', async () => {
      const badState = {
        email: 'trashman@gmail.com',
        password: 'ismellbad'
      }

      wrapper.setState(badState);

      const result = await wrapper.instance().verifyUser()
      const expected = undefined;

      expect(result).toEqual(expected);
    })
  })

  describe('handleSubmit', () => {

    it('calls verifyUser with the correct arguments', () => {

    })

    it('calls alert if verifyUser returns undefined', () => {

    })

    it('calls props.updateCurrentUser with the correct arguments if user has been verified', () => {

    })

    it('calls getFavorites with correct arguments if user has been verified', () => {

    })

    it('calls props.updateFavorites with the correct arguments if user has been verfied', () => {

    })

    it()
  })
  
});
