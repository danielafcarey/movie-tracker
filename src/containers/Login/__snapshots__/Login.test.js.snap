// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    updateCurrentUser={[Function]}
    updateFavorites={[Function]}
    updateMovies={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          name="email"
          onChange={[Function]}
          placeholder="Email"
          required={true}
          type="text"
          value=""
/>,
        <input
          name="password"
          onChange={[Function]}
          placeholder="Password"
          required={true}
          type="password"
          value=""
/>,
        <p />,
        <button>
          Log In
</button>,
      ],
      "onSubmit": [Function],
      "type": "submit",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "email",
          "onChange": [Function],
          "placeholder": "Email",
          "required": true,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "password",
          "onChange": [Function],
          "placeholder": "Password",
          "required": true,
          "type": "password",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
        },
        "ref": null,
        "rendered": "",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Log In",
        },
        "ref": null,
        "rendered": "Log In",
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            name="email"
            onChange={[Function]}
            placeholder="Email"
            required={true}
            type="text"
            value=""
/>,
          <input
            name="password"
            onChange={[Function]}
            placeholder="Password"
            required={true}
            type="password"
            value=""
/>,
          <p />,
          <button>
            Log In
</button>,
        ],
        "onSubmit": [Function],
        "type": "submit",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "email",
            "onChange": [Function],
            "placeholder": "Email",
            "required": true,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "password",
            "onChange": [Function],
            "placeholder": "Password",
            "required": true,
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Log In",
          },
          "ref": null,
          "rendered": "Log In",
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
